### PR TITLE
Adds a way to preserve user-preferred mesh grid state

### PR DESCRIPTION
**Pull request by mockoocy**
Fixes a bug / bad UX in which caused mesh grid to became unhidden after certain operations like DataCollection or Phase change.

The steps to reproduce the bugs were as following:
- mount a sample
- open `draw_grid` control
- draw a few grids
- make some of them hidden
- right click on canvas, press `Data Collection` button
- run the task
After the task was done, each of the grids was shown again.
It's also worth adding that in the demo beamline many of the motors seem to not be mocked, so the behavior can't be observed. 

The reason for this behavior was further described in this Pull request https://github.com/mxcube/mxcubecore/pull/1044 for mxcubecore. This PR will be marked as draft until the mxcubecore one is merged.

The fix writes to state set by user to `Shape.user_state` attribute on the mxcubecore side.
When the data collection or phase change is finished, then the user-preferred state is restored by setting `Shape.state` to be equal to `Shape.user_state`.

### CODE REVIEW

**Comment by mockoocy** [comment]
Turns out the same issue happens for the zoom and focus controls. Will also try to fix it for these in this PR.

**Comment by fabcor-maxiv** [comment]
> Turns out the same issue happens for the zoom and focus controls. Will also try to fix it for these in this PR.

@mockoocy Note that there was work on focus and zoom controls very recently: <https://github.com/mxcube/mxcubeweb/pull/1457>. I do not know if it is related or not.

**Comment by mockoocy** [comment]
added some changes; 
now the UI has some indicator that specific grid is being hidden (it appears with a darker background)
I have also changed `HWR.beamline.sample_view.add_shape_from_mpos` function, also when updating shapes there are optionally more arguments passed to it, so that `shape` and `user_shape` are not changed when grids is being redrawn sometimes.

**Comment by marcus-oscarsson** [comment]
Thanks @mockoocy, do you happen to have a screenshot of the changes ?

**Comment by mockoocy** [comment]
> Thanks @mockoocy, do you happen to have a screenshot of the changes ?

image

File: ui/src/components/SampleView/SampleView.css [hunk 1]
```diff
@@ -176,3 +176,7 @@
   color: #000;
   font-weight: bold;
 }
+
+.hid-by-user {
+  background-color: #aaa !important;
```
Comment (fabcor-maxiv) [inline ui/src/components/SampleView/SampleView.css]: Isn't it usually bad practice to use `!important` in CSS?

**Comment by elmjag** [comment]
@marcus-oscarsson I think we have different ideas how this mesh hidding/un-hidding is suppose to work.

Like what is the actual desired behavior. Perhaps we can discuss this in a video meeting? I'm at least a bit confused right now.

File: ui/src/components/SampleView/SampleView.css (hunk 1 above)
Comment (mockoocy) [reply ui/src/components/SampleView/SampleView.css]: You're right. I switched to using data attributes to override bootstrap's default css.

**Comment by marcus-oscarsson** [comment]
@elmjag sure, we can have a quick chat if you like. How about tomorrow ? Otherwise I think the current proposal is fine maybe the color could be a bit more fun (other than dark gray :) ) and what do you think about a text perhaps under the hide button (saying "hidden by user" and perhaps "out of view" when hidden automatically) ?

**Comment by mockoocy** [comment]
Undrafting.
As per @marcus-oscarsson's suggestion I changed the background color for table with grid info. 
Additionally I added a tooltip that shows the reason why a grid is hidden. I think it's nicer and informative enough :) 

image
withTooltip

**Comment by marcus-oscarsson** [comment]
Would it be possible to get some more distinct color difference, between "hidden by user" and "out of range", like light blue and light green or something ?

**Comment by marcus-oscarsson** [comment]
Or maybe no change for "our of range" and light blue or green for "hidden by user" ?

**Comment by mockoocy** [comment]
> Or maybe no change for "our of range" and light blue or green for "hidden by user" ?

Light green's already reserved for selected grid, so I guess we'll go with light blue  :smile_cat:   
image

**Comment by marcus-oscarsson** [comment]
Ok, yeah that looks better I guess, does it not ?. Would it make sense to make the text of the hidden meshes italic ?

**Comment by mockoocy** [comment]
> Ok, yeah that looks better I guess, does it not ?. Would it make sense to make the text of the hidden meshes italic ?

Yeah, that's a good idea - done :)

**Comment by marcus-oscarsson** [comment]
Cool, thanks alot very nice !

File: ui/src/components/SampleView/GridForm.jsx [hunk 2]
```diff
@@ -5,10 +5,27 @@ import Draggable from 'react-draggable';
...
 import { useSelector } from 'react-redux';
+import TooltipTrigger from '../TooltipTrigger';
 
 function handleContextMenu(e) {
   e.stopPropagation();
 }
+/**
+ * Returns the tooltip content for the hide button
+ *
+ * @param {boolean} userHidden - if the grid is hidden by the user
```
Comment (marcus-oscarsson) [inline ui/src/components/SampleView/GridForm.jsx:16]: :+1: Nice, thanks

File: ui/src/components/SampleView/SampleView.css [hunk 3]
```diff
@@ -164,8 +164,13 @@
...
 
-.selected {
-  background-color: #dff0d8 !important;
+table.table-striped > tbody > tr[data-selected='true'] {
+  background-color: #dff0d8;
   color: #000;
   font-weight: bold;
 }
+
+table.table-striped > tbody > tr[data-user-hidden='true'] {
```
Comment (marcus-oscarsson) [inline ui/src/components/SampleView/SampleView.css:173]: :+1:

**Comment by marcus-oscarsson** [comment]
I guess this is good to go then ?, Ill merge  https://github.com/mxcube/mxcubecore/pull/1044 then you can update `pyrproject.toml`

**Comment by marcus-oscarsson** [comment]
My apologies it took sometime to get PR on `mxcubecore` merged, its done now :)

**Pull request merged**